mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -830,17 +830,617 @@ LogicalResult ThreadwiseWriteAllRewritePattern::matchAndRewrite(
+    TransformMapAttr padTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, padTrMap);
+    threadsToTensor.unmerge({"tid", "nrIter"}, {0, 1}, "nrDim",
+                            {blockSize, nrThreads});
+    threadsToTensor.passThrough({"rIter"}, {2}, {"rIter"});
+    TransformMapAttr unmergeTrMap = threadsToTensor.get();
+
+    return rewriter.getArrayAttr({unmergeTrMap, padTrMap, mergeTrMap});
+  }
+
+  // This should only be used if product non-reduction dims is
+  // less than number threads in a block.
+  //
+  // Given a input tensor : D0, ... , Dr , ... , DN to reduce,
+  // This function creates a view that maps the space of
+  // [D0, ... , Dr , ... , DN] --> [nrtid, rtid, rIter] where
+  // nrtid = tid / product(non-reduction dims) is a reduction subgroup leader.
+  // rtid = tid % product(non-reduction dims) is thread idx within a reduction
+  // subgroup. | rtid | is the number of threads that'd participate in a
+  // reduction
+  ArrayAttr createThreadViewforNRSmallerThanThreads(
+      Location loc, ArrayRef<int64_t> toReduceShape, int64_t blockSize,
+      size_t reduceAxis, PatternRewriter &rewriter) const {
+    BottomUpTMBuilder threadsToTensor(rewriter, toReduceShape, loc);
+    SmallVector<StringRef, 4> lowerNameRefs;
+    threadsToTensor.getStartNames(lowerNameRefs);
+    int64_t nonReduceMergeDimSize = 1;
+    SmallVector<StringRef, 4> nonReduceNameRefs;
+    for (auto [dim, dimSize] : llvm::enumerate(toReduceShape)) {
+      if (dim != reduceAxis) {
+        nonReduceMergeDimSize *= dimSize;
+        nonReduceNameRefs.push_back(lowerNameRefs[dim]);
+      }
+    }
+    threadsToTensor.merge("nrDim", 0, nonReduceNameRefs);
+    threadsToTensor.passThrough({"rDim"}, {1}, {lowerNameRefs[reduceAxis]});
+    TransformMapAttr mergeTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, mergeTrMap);
+    // Distribute threads that is extra than nonReduceMergeDimSize and use them
+    // for reductions. This is a floor because max threads per non-reduction
+    // axes and we use remaining for reductions.
+    int64_t rthreads = blockSize / nonReduceMergeDimSize;
+    int64_t rDimPerRThread =
+        (toReduceShape[reduceAxis] + (rthreads - 1)) / rthreads;
+    threadsToTensor.pad(
+        {"rDim"}, {0, rthreads * rDimPerRThread - toReduceShape[reduceAxis]});
+    threadsToTensor.passThrough({"nrDim"}, {0}, {"nrDim"});
+    TransformMapAttr padTrMap = threadsToTensor.get();
+
+    threadsToTensor = BottomUpTMBuilder::above(threadsToTensor, padTrMap);
+    threadsToTensor.unmerge({"rtid", "rIter"}, {1, 2}, "rDim",
+                            {rthreads, rDimPerRThread});
+    threadsToTensor.passThrough({"nrtid"}, {0}, {"nrDim"});
+    TransformMapAttr unmergeTrMap = threadsToTensor.get();
+
+    return rewriter.getArrayAttr({unmergeTrMap, padTrMap, mergeTrMap});
+  }
+
+  ArrayAttr dropDim(Location loc, ArrayAttr view, size_t dropDimIdx,
```

Comment:
What's this for?

Reply:
This to drop the bid dim from the view. Please see the reply for having gridsize here..